Created recipe endpoints

Creepsy opened the pull request:
- "/recipes" list all recipes
- "/recipes/:id" get the specified recipe
- post request to "/recipes" to add a new recipe

rehwinkel committed:
implemented registration and login

Creepsy committed:
Created recipe endpoints
- "/recipes" list all recipes
- "/recipes/:id" get the specified recipe
- post request to "/recipes" to add a new recipe

applied requested patches of merge request

rehwinkel commented on src/Main.hs:
```diff
@@ -25,6 +27,11 @@ import Database.SQLite.Simple.ToField (ToField (toField))
 import GHC.Generics (Generic)
 import Network.HTTP.Types (status201, status401, status409)
+import Data.Text hiding (map)
+import Database.SQLite.Simple
+import Database.SQLite.Simple.ToField (ToField (toField))
+import GHC.Generics (Generic)
+import Network.HTTP.Types (status401, status404)
```
Bitte duplicates entfernen und status404 oben hinzufügen

Creepsy replied:
Should get resolved by 9d8ac164eaaf6e8355e1254ae071ac5eb29b6aef

rehwinkel commented on src/Main.hs:
```diff
@@ -53,6 +60,26 @@ createTables :: Connection -> IO ()
+  execute_
+    db
+    "CREATE TABLE IF NOT EXISTS recipes (\
+    \ id INTEGER PRIMARY KEY AUTOINCREMENT,\
```
vlt `recipeId` nennen

Creepsy replied:
Renamed in 9d8ac164eaaf6e8355e1254ae071ac5eb29b6aef

rehwinkel commented on src/Main.hs:
```diff
@@ -147,6 +174,73 @@ checkSession db = do
         Just session -> return session
         Nothing -> raiseStatus status401 "Invalid session cookie"
     Nothing -> raiseStatus status401 "Invalid session cookie"
+data Ingredient = Ingredient {count :: Int, ingredient :: String} deriving (Read, Show, Generic)
```
`count` ist nicht immer angemessen. Hier sollte in einem zukünftigen MR verschiedene Units supported werden (grams, mililiters, none)

Creepsy replied:
Not a part of this merge request, but rather a feature that can be implemented in the future. Not resolved by 9d8ac164eaaf6e8355e1254ae071ac5eb29b6aef

rehwinkel commented on src/Main.hs:
```diff
@@ -147,6 +174,73 @@ checkSession db = do
+      ]
+
+getAllRecipes :: Connection -> IO [RecipeWithID]
+getAllRecipes = flip query_ "SELECT * FROM recipes"
```
semicolon in query fehlt

Creepsy replied:
Resolved in 9d8ac164eaaf6e8355e1254ae071ac5eb29b6aef

rehwinkel commented on src/Main.hs:
```diff
@@ -147,6 +174,73 @@ checkSession db = do
+      "INSERT INTO recipes (title, description, ingredients, timeMinutes, costCents, stars) VALUES (?, ?, ?, ?, ?, ?);"
+      $ toRow recipe
+    return True
+    `catch` (\(_ :: FormatError) -> return False)
```
Dont catch! Dieser fehler darf nie eintreten, also muss der catch weg (dann bitte auch das language feature dazu entfernen)

Creepsy replied:
Catch and ScopedTypeVariables removed in 9d8ac164eaaf6e8355e1254ae071ac5eb29b6aef

rehwinkel commented on src/Main.hs:
```diff
@@ -147,6 +174,73 @@ checkSession db = do
+    `catch` (\(_ :: FormatError) -> return False)
+
+getRecipe :: Connection -> RecipeID -> IO (Occurence Recipe)
+getRecipe db recipeID = occurences . map recipe <$> query db "SELECT * FROM recipes WHERE id = ?" (Only recipeID)
```
Fehlt semicolon.

Creepsy replied:
Resolved in 9d8ac164eaaf6e8355e1254ae071ac5eb29b6aef

rehwinkel commented on src/Main.hs:
```diff
@@ -180,3 +274,17 @@ main = do
+      recipe <- jsonData :: ActionM Recipe
+      success <- liftIO $ addRecipe db recipe
+      if success
+        then html "Recipe inserted into database!"
```
Hier bitte status "201 created" nutzen

Creepsy replied:
9d8ac164eaaf6e8355e1254ae071ac5eb29b6aef now uses status 201 instead of html

rehwinkel commented on src/Main.hs:
```diff
@@ -180,3 +274,17 @@ main = do
+      success <- liftIO $ addRecipe db recipe
+      if success
+        then html "Recipe inserted into database!"
+        else raiseStatus status401 "Error occured on recipe insert!"
```
Dieser fall tritt nie ein. Falls doch, sollte der server crashen.

Creepsy replied:
9d8ac164eaaf6e8355e1254ae071ac5eb29b6aef doesn't handle this exception anymore. Changed addRecipe's return type to () as it should never fail.

Creepsy committed:
moved recipe endpoints into extra file

Creepsy commented:
Rebased branch with newest commits from main branch and moved recipe endpoint into an extra file.
Note: The current recipe endpoint **doesn't** use the new ingredients endpoint/API.

rehwinkel commented on src/Recipe.hs:
```diff
@@ -0,0 +1,132 @@
+    handleAddRecipe,
+    getAllRecipes,
+    getRecipe,
+    addRecipe,
```
es sollten eigentlich nur die handler und DB tables exportiert werden.

Creepsy replied:
Exports of "getAllRecipes, getRecipe, addRecipe" removed in 61808d2e21d33e8bbea750823776691997934ac5

rehwinkel commented on src/Main.hs:
```diff
@@ -1,7 +1,6 @@
-{-# LANGUAGE DuplicateRecordFields #-}
-{-# LANGUAGE ImportQualifiedPost #-}
 {-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE ImportQualifiedPost #-}
```
Kein grund das zu ändern

rehwinkel commented on src/Main.hs:
```diff
@@ -10,18 +9,24 @@ import Data.Aeson (decode)
+
+data Color = Color {peter :: String, jonas :: Int} deriving (Generic)
+
+instance ToJSON Color
```
Das hab ich bewusst gelöscht, bitte nicht wieder hinzufügen.

Creepsy replied:
Removed in 61808d2e21d33e8bbea750823776691997934ac5

Creepsy committed:
resolved issues of merge request
